mesonbuild/modules/_qt.py

```diff
@@ -616,3 +734,364 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
             return ModuleReturnValue(results.return_value[0], [results.new_objects, translations])
         else:
             return ModuleReturnValue(translations, [translations])
+
+    def _source_abs_path(self, state: 'ModuleState', source: T.Union[FileOrString, build.CustomTarget, build.CustomTargetIndex]) -> str:
+        if isinstance(source, str):
+            return os.path.join(state.environment.source_dir, state.subdir, source)
+        elif isinstance(source, File):
+            return source.absolute_path(state.environment.source_dir, state.environment.build_dir)
+        elif isinstance(source, (build.CustomTarget, build.CustomTargetIndex)):
+            output: str = source.get_filename()
+            return os.path.join(state.environment.get_build_dir(), state.backend.get_target_dir(source), output)
+        else:
+            assert False, "unreachable"
```

Comment:
I'm happy to trust the type annotations for the final release, though this might be useful for bringup

Reply:
I changed this function quite a bit, I took inspiration from the one in gnome.py